### PR TITLE
Improves the LuaJIT performance by using luaL_openlibs to initialize the VM

### PR DESCRIPTION
**Pull request by modlfo**
As I mentioned in my other pull request. This change doubles the performance of my complex Lua code. In order to forbid the unsafe libraries, like `io` and `os`, they are set to nil just after initializing the VM.

I also set to `nil` the function `require` that way no other libraries can be loaded.

### CODE REVIEW

File: src/LuaJITEngine.cpp [hunk 1]
```diff
@@ -28,12 +28,16 @@ struct LuaJITEngine : ScriptEngine {
...
-		luaopen_table(L);
-		luaopen_math(L);
-		luaopen_bit(L);
+		// Import the common Lua libraries
+		luaL_openlibs(L);
+		// Unloads the unsafe libraries
+		lua_pushinteger(L, 0);
+		lua_setglobal(L, "os");
+		lua_pushinteger(L, 0);
+		lua_setglobal(L, "io");
```
Comment (AndrewBelt) [inline src/LuaJITEngine.cpp:37]: What about the package, debug, jib, and ffi packages? Are those not loaded with `luaL_openlibs`?

File: src/LuaJITEngine.cpp [hunk 2]
```diff
@@ -28,12 +28,16 @@ struct LuaJITEngine : ScriptEngine {
...
-		// Import a subset of the standard library
-		luaopen_base(L);
-		luaopen_string(L);
-		luaopen_table(L);
-		luaopen_math(L);
-		luaopen_bit(L);
+		// Import the common Lua libraries
+		luaL_openlibs(L);
+		// Unloads the unsafe libraries
+		lua_pushinteger(L, 0);
```
Comment (AndrewBelt) [inline src/LuaJITEngine.cpp:34]: Could you push nil instead?

**Comment by modlfo** [comment]
I found the reason for the speedup. The JIT library needs to be loaded in order to turn the JIT engine on, otherwise it will be off. I have loaded the library and then set it to nil so it cannot be used in the script.

**Comment by AndrewBelt** [comment]
Hahaha, I didn't even have JIT enabled??

**Comment by teknico** [comment]
> Hahaha, I didn't even have JIT enabled??

It did seem to run slower than Javascript when I tried it.

**Comment by modlfo** [comment]
There's still something strange. In my benchmarks, LuaJIT most of the times performs almost as fast as C++. The main case when LuaJIT is slower is when the program has large arrays. The main difference, in the case of my benchmarks, is that the LuaJIT executable runs the whole tests (it's not embedded in other executable).

I will keep investigating when I have some time.

**Comment by modlfo** [comment]
I have the sensation that the performance hit is in the reading and writing of the `block` information. Some years ago I was working on my AudioEngine based on LuaJIT and seems like I used the LuaJIT FFI to have fast access to the audio buffers.

https://github.com/modlfo/AudioEngine/blob/master/AudioEngine/LuaScripts/processor.lua

I will try to make a version that uses the FFI to access directly the `ProcessBlock`.

**Comment by JerrySievert** [comment]
> I have the sensation that the performance hit is in the reading and writing of the block information. Some years ago I was working on my AudioEngine based on LuaJIT and seems like I used the LuaJIT FFI to have fast access to the audio buffers.

that is typically the issue with javascript engines as well.  the c++/interpreter membrane ends up being the expensive part of the equation.

**Comment by modlfo** [comment]
Using the LuaJIT FFI to access the `ProcessBlock` my plugin went from consuming 25% to 5%. Just for reference, the C++ version consumes 2%. 

I will prepare a new pull request for this improvement.

**Comment by AndrewBelt** [comment]
For turning JIT on, is `luaJIT_setmode` the function you want? Probably `luaopen_jit` and `luaL_openlibs` call this function, but if we don't want the `jit.` library, maybe calling `luaJIT_setmode` manually is the best way to do it.

**Pull request merged**